test/distributed/checkpoint/test_pg_aware_planner.py

```diff
@@ -0,0 +1,185 @@
+    FullyShardedDataParallel as FSDP,
+)
+from torch.distributed.fsdp.fully_sharded_data_parallel import StateDictType
+from torch.testing._internal.common_utils import run_tests
+from torch.distributed.checkpoint.pg_aware_planner import (
+    ProcessGroupAwareSavePlanner,
+    ProcessGroupAwareLoadPlanner,
+)
+from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    with_comms,
+    skip_unless_torch_gpu,
+)
+
+
+class MyModule(torch.nn.Module):
+    def __init__(
+        self, rank: int, extra_state: int, extra_state_tensor: torch.Tensor
+    ) -> None:
+        super().__init__()
+        self.param = torch.nn.Parameter(
+            torch.arange(
+                start=rank * 4, end=rank * 4 + 4, step=1, dtype=torch.float32
+            )
+        )
+        self._extra_state = extra_state
+        self._extra_state_tensor = extra_state_tensor
+
+    @property
+    def extra_state(self) -> int:
+        return self._extra_state
+
+    @extra_state.setter
+    def extra_state(self, new_extra_state: int) -> None:
+        self._extra_state = new_extra_state
+
+    @property
+    def extra_state_tensor(self) -> torch.Tensor:
+        return self._extra_state_tensor
+
+    @extra_state_tensor.setter
+    def extra_state_tensor(self, new_extra_state_tensor: torch.Tensor) -> None:
+        self._extra_state_tensor = new_extra_state_tensor
+
+    def get_extra_state(self) -> Dict[str, Union[int, torch._tensor.Tensor]]:
+        return {
+            "extra_state": self._extra_state,
+            "extra_state_tensor": self._extra_state_tensor,
+        }
+
+    def set_extra_state(
+        self, state: Dict[str, Union[int, torch._tensor.Tensor]]
+    ) -> None:
+        self._extra_state = state["extra_state"]  # pyre-ignore[8]
+        self._extra_state_tensor = state["extra_state_tensor"]  # pyre-ignore[8]
+
+
+class TestProcessGroupAwarePlanner(DTensorTestBase):
+    def _create_new_dist_group(self):
+        world_size = dist.get_world_size()
+        group1 = [i for i in range(world_size) if i % 2 == 0]
+        group2 = [i for i in range(world_size) if i % 2 != 0]
+
+        # create new fsdp group for resharding
+        fsdp_0 = dist.new_group(ranks=group1)
+        fsdp_1 = dist.new_group(ranks=group2)
+        if dist.get_rank() % 2 == 0:
+            my_fsdp = fsdp_0
+        else:
+            my_fsdp = fsdp_1
+
+        return my_fsdp
+
+
+    @with_comms
+    @skip_unless_torch_gpu
+    @with_temp_dir
+    def test_process_group_aware_planner(self) -> None:
+        CHECKPOINT_DIR = self.temp_dir
+
+        model = MyModule(
+            rank=dist.get_rank(),
+            extra_state=0,
+            extra_state_tensor=torch.tensor([[1.0, -1.0], [1.0, -1.0]]),
+        ).cuda(dist.get_rank())
+
+        model = FSDP(model, process_group=self._create_new_dist_group())
```

Comment:
We should explain that by using 2 PGs we're creating two in-memory models in the cluster whose only difference is the PG they got assigned.

Reply:
Sounds good. Will update in a separate PR.